mistletoe/block_token.py

```diff
@@ -764,40 +766,57 @@ def read(cls, lines):
 
     @classmethod
     def match_reference(cls, lines, string, offset):
+        # up to three spaces, "[", label, "]"
         match_info = cls.match_link_label(string, offset)
         if not match_info:
-            cls.backtrack(lines, string, offset)
             return None
         _, label_end, label = match_info
 
+        # ":"
         if not follows(string, label_end-1, ':'):
-            cls.backtrack(lines, string, offset)
             return None
 
-        match_info = cls.match_link_dest(string, label_end)
+        # optional spaces or tabs (including up to one line ending)
+        dest_start = shift_whitespace(string, label_end + 1)
+        if dest_start == len(string):
+            return None
+
+        # link destination
+        match_info = cls.match_link_dest(string, dest_start)
         if not match_info:
-            cls.backtrack(lines, string, offset)
             return None
         _, dest_end, dest = match_info
 
-        match_info = cls.match_link_title(string, dest_end)
-        if not match_info:
-            cls.backtrack(lines, string, dest_end)
+        # optional spaces or tabs (including up to one line ending)
+        title_start = shift_whitespace(string, dest_end)
+        if title_start == dest_end and title_start < len(string):
             return None
+
+        # optional link title, which if it is present must be separated from the link
+        # destination by spaces or tabs
+        match_info = cls.match_link_title(string, title_start)
+        if not match_info:
+            eol = string[dest_end:title_start].find("\n")
+            if eol >= 0:
+                return dest_end + eol + 1, (label, dest, "")
+            else:
+                return None
```

Comment:
Adding some more explanatory comments and modifying the naming a little to make it more clear what's going on.

Yeah I wasn't entirely happy with the code duplication. But on the other hand, it's just four lines of code. I considered breaking them out into a separate function, but it didn't make it easier to read.

And no, that change to `match_link_title` won't help, because the newline that we're looking for needs to be in the whitespace *between* the destination and the title.

Reply:
OK, better now, thank you.